internal/prepare/prepare_test.go

```diff
@@ -0,0 +1,67 @@
+package prepare_test
+
+import (
+	"testing"
+
+	"github.com/canonical/sqlair/internal/parse"
+	"github.com/canonical/sqlair/internal/prepare"
+	. "gopkg.in/check.v1"
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) { TestingT(t) }
+
+type ParserSuite struct{}
+
+var _ = Suite(&ParserSuite{})
+
+type Address struct {
+	ID       int    `db:"id"`
+	District string `db:"district"`
+	Street   string `db:"street"`
+}
+
+type Person struct {
+	ID         int    `db:"id"`
+	Fullname   string `db:"name"`
+	PostalCode int    `db:"address_id"`
+}
+
+func (s *ParserSuite) TestValidPrepare(c *C) {
+	testList := []struct {
+		input            string
+		prepareArgs      []any
+		expectedPrepared string
+	}{{
+		"SELECT street FROM t WHERE x = $Address.street",
+		[]any{Address{}},
+		"SELECT street FROM t WHERE x = ?",
+	}, {
+		"SELECT p FROM t WHERE x = $Person.id",
+		[]any{Person{}},
+		"SELECT p FROM t WHERE x = ?",
+	}}
+	for _, test := range testList {
+		parser := parse.NewParser()
+		parsedExpr, _ := parser.Parse(test.input)
+		preparedExpr, err := prepare.Prepare(parsedExpr, test.prepareArgs...)
+		c.Log(err)
+		c.Assert(preparedExpr.SQL, Equals, test.expectedPrepared)
+	}
+}
+
+func (s *ParserSuite) TestMismatchedInputStructName(c *C) {
+	sql := "SELECT street FROM t WHERE x = $Address.street"
+	parser := parse.NewParser()
+	parsedExpr, err := parser.Parse(sql)
+	_, err = prepare.Prepare(parsedExpr, Person{ID: 1})
+	c.Assert(err, ErrorMatches, "cannot prepare expression: unknown type: Address")
```

Comment:
I may not have understood the test here, so please ignore if not relevant :) In line, 57, we deliberately pass a `Person` to `Prepare()`, so should it complain about the `sql` statement having an Address or about the second argument being a `Person`? Does this test have the expectation that `Prepare()` must complain when it receives an expression that does not contain the provided type (`Person`, in this case). I guess the error is produced by `prepareInput()`. Could that method equivalent complain about the provided argument type itself? or do we always blame the sql expression :P

Reply:
Prepare basically takes all the type arguments it gets into a bag and looks in the bag for the ones it finds in the query. So if extra arguments are given then they are ignored.

Admittedly this behaviour isn't really documented, so I will add that!